### ISSUE TITLE
Wrong Serialization (CompatibleFieldSerializer)

### ISSUE THREAD

**Issue by oassuncao** [EsotericSoftware/kryo#556]
I use kryo to serialize information between my systems, but I've some problems when some fields not exists in the class when I read

I created a project to simulate the problem

I created a class named `DefaultClassResolver` only to change the Class Types and simulate other system and other classes with the same structure

I have the original classes in the package `kryo.test.SupplierInitial` and the classes of other system in the package `kryo.test.SupplierWrapper`

If run the app without any change the output is **"Wrong"**, because the **nature** field receive the value of the **link** field, this occurs because the class `SupplierAggregationME` in the package **kryo.test.SupplierInitial** has 2 fields (value and total) and in the class `SupplierAggregationME` of package **kryo.test.SupplierWrapper** doesn't have.

If you uncomment the fields and getter and setter, the problem doesn't occur, The output is **"Correct"**

This is only an example, if you comment the fields (getter and setter) on the SupplierME classes (SupplierInitial and SupplierWrapper packages) you can verify the **nature** fields receive the value of **name** field

[kryotest.zip](https://github.com/EsotericSoftware/kryo/files/1520096/kryotest.zip)

**Comment by oassuncao** [comment]
Anybody has any information about this bug?

**Comment by oassuncao** [comment]
Hey guys 

Someone can you help me?

**Comment by magro** [comment]
@oassuncao It would help if you could further reduce the sample, following the ideas of http://www.sscce.org.

**Comment by oassuncao** [comment]
@magro This is a reduced example of the original code, I could only reproduce this error with this composition classes

If I remove a field of the POJO, the error does not occurs more

Please try to help me, I have this problem in the production environment

**Comment by magro** [comment]
Can you remove as many as classes as possible to see where/when the error starts to occur?

**Comment by oassuncao** [comment]
I removed all classes as possible, If I remove any other class, the error does not occur

[kryotest_slim.zip](https://github.com/EsotericSoftware/kryo/files/1831667/kryotest_slim.zip)

**Comment by magro** [comment]
Ok, I analyzed this and could track it down to the `NatureType`. If you change `DocumentType.type` from `NatureType` to `String` it works. I also removed the differences between the two `NatureType`s, unfortunately this did not solve it. That's everything I could find so far, I'll see if I can dig deeper.

**Comment by oassuncao** [comment]
Thank you @magro for you support

Check if you remove the "aggregations" data on data.json, the problem not occur more

```json
{
  "suppliers": [
    {
      "nature": "LegalEntity",
      "documents": [
        {
          "value": "42216103000133",
          "type": {
            "type": "LegalEntity"
          }
        }
      ]
    }
  ]
}
```

**Comment by magro** [comment]
Yes, it's caused by the `LegalEntity` which occurs twice. There seems to be an issue with the reference handling - i.e. if you add `kryo.setReferences(false);` it works!

**Comment by magro** [comment]
More findings: because the `SupplierAggregationME.value` field (with value "Brazil") is missing on read side its object reference is not read and not added to `Kryo.readReferenceIds`, so that in consequence the following reference ids are off by one. Because during write the second `LegalEntity` was stored as object reference to the first one, the object registered for the reference id is used - which is then `42216103000133` instead of `LegalEntity` (off by one).

**Comment by magro** [comment]
@oassuncao D'oh, after all that digging I just found that this issue had actually been analyzed and discussed in the past - see https://github.com/EsotericSoftware/kryo/issues/286#issuecomment-74870545. And in fact it's also documented in [CompatibleFieldSerializer](https://github.com/EsotericSoftware/kryo/blob/master/src/com/esotericsoftware/kryo/serializers/CompatibleFieldSerializer.java#L38).

If you follow the discussion and links you'll see that this is not easily being solved. Can you for now disable referenced to get this solved? Not sure if / when there's a solution for this.
